services/upgrader/upgrader.go

```diff
@@ -0,0 +1,372 @@
+		upgradePercentage = 0
+		return err
+	}
+	defer os.Remove(filename)
+	sigfilename, err := download(getCurrentVersionTgzUrl()+".sig", 1024)
+	if err != nil {
+		if sigfilename != "" {
+			os.Remove(sigfilename)
+		}
+		upgradeError = errors.New("error downloading the new version signature file")
+		mlog.Error("Unable to download the mattermost server version signature file", mlog.String("url", getCurrentVersionTgzUrl()+".sig"), mlog.Err(err))
+		upgradePercentage = 0
+		return err
+	}
+	defer os.Remove(sigfilename)
+
+	err = verifySignature(filename, sigfilename, mattermostBuildPublicKey)
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = errors.New("Unable to verify downloaded file signature")
+		mlog.Error("Unable to verify downloaded file signature", mlog.Err(err))
+		return err
+	}
+
+	err = extractBinary(executablePath, filename)
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = err
+		mlog.Error("Unable to extract the binary from the downloaded file", mlog.Err(err))
+		return err
+	}
+	upgradePercentage = 100
+	return nil
+}
+
+func UpgradeToE0Status() (int64, error) {
+	return upgradePercentage, upgradeError
+}
+
+func download(url string, limit int64) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	out, err := ioutil.TempFile("", "*_mattermost.tar.gz")
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+
+	counter := &writeCounter{total: resp.ContentLength}
+	_, err = io.Copy(out, io.TeeReader(&io.LimitedReader{R: resp.Body, N: limit}, counter))
+	return out.Name(), err
+}
+
+func getFilePermissionsOrDefault(filename string, def os.FileMode) os.FileMode {
+	file, err := os.Open(filename)
+	if err != nil {
+		mlog.Error("Unable to get the file permissions", mlog.String("filename", filename), mlog.Err(err))
+		return def
+	}
+	fileStats, err := file.Stat()
+	if err != nil {
+		mlog.Error("Unable to get the file permissions", mlog.String("filename", filename), mlog.Err(err))
+		return def
+	}
+	return fileStats.Mode()
+}
+
+func extractBinary(executablePath string, filename string) error {
+	gzipStream, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+
+	uncompressedStream, err := gzip.NewReader(gzipStream)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(uncompressedStream)
+
+	for {
+		header, err := tarReader.Next()
+
+		if err == io.EOF {
+			return errors.New("Unable to find mattermost binary in the downloaded version")
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if header.Typeflag == tar.TypeReg && header.Name == "mattermost/bin/mattermost" {
+			permissions := getFilePermissionsOrDefault(executablePath, 0755)
+			tmpFile, err := ioutil.TempFile("", "*")
+			if err != nil {
+				return err
+			}
+			tmpFileName := tmpFile.Name()
+			os.Remove(tmpFileName)
+			err = os.Rename(executablePath, tmpFileName)
```

Comment:
This looks like a race. Not sure how winnable, but nevertheless. Consider the following:

 - An attacker monitors `/tmp` as the unprivileged user `evil`
 - The upgrader creates `tmpFile` and sets `tmpFileName` to `/tmp/6b2453a7-4351-4c73-8d6d-4614c9711c4d`
 - Immediately after the `os.Remove` call, `evil` creates a *world-writable directory* at `/tmp/6b2453a7-4351-4c73-8d6d-4614c9711c4d`
 - The upgrader proceeds to the `os.Rename` call, which succeeds because instead of renaming the file it ends up moving the file to the attacker-created directory
 - Now `evil` controls the backup and if the upgrade fails, a malicious copy may be restored

Reply:
@jupenur changing it to use a temp file in the same mattermost directory instead of the tmp directory would address your concerns?

Reply:
It would, yeah. Another alternative would be to use `ioutl.TempDir` instead of `ioutil.TempFile` so that the `os.Remove` call can be done away with.